parquet-common/src/main/java/org/apache/parquet/bytes/SequenceByteBufferInputStream.java

```diff
@@ -0,0 +1,269 @@
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package org.apache.parquet.bytes;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ *   A bare minimum implementation of a {@link java.io.SequenceInputStream} that wraps an
+ *   <i>ordered</i> collection of ByteBufferInputStreams.
+ *   <p>
+ *   This class, as implemented, is intended only for a specific use in the ParquetFileReader and
+ *   throws {@link UnsupportedOperationException} in unimplemented methods to catch any unintended
+ *   use in other cases.
+ *   <p>
+ *   Even thought this class is derived from ByteBufferInputStream it explicitly does not support any
+ *   byte buffer related methods like slice. It does, however support sliceBuffers which is a
+ *   curious case of reading data from underlying streams
+ *   <p>
+ *   Even though this class changes the state of the underlying streams (by reading from them)
+ *   it does not own them and so the close method does not close the streams. To avoid resource
+ *   leaks the calling code should close the underlying streams
+ */
+public class SequenceByteBufferInputStream extends ByteBufferInputStream {
+
+  Collection<ByteBufferInputStream> collection;
+  Iterator<ByteBufferInputStream> iterator;
+  ByteBufferInputStream current;
+  long position = 0;
+
+  @Override
+  public String toString() {
+    return "SequenceByteBufferInputStream{" +
+      "collection=" + collection +
+      ", current=" + current +
+      ", position=" + position +
+      '}';
+  }
+
+  public SequenceByteBufferInputStream(Collection<ByteBufferInputStream> collection) {
+    this.collection = collection;
+    iterator = collection.iterator();
+    current = iterator.hasNext() ? iterator.next() : null;
+    if (current == null) {
+      throw new UnsupportedOperationException(
+        "Initializing SequenceByteBufferInputStream with an empty collection is not supported");
+    }
+  }
+
+  @Override
+  public long position() {
+    return position;
+  }
+
+  @Override
+  public int read(ByteBuffer out) {
```

Comment:
are you 100% confident that all uses of read() in parquet are in a single thread? it is not unknown for apps to read across threads, even though the java APIs say "don't"

Reply:
good to know.